src/Client/Updater.php

```diff
@@ -575,4 +525,70 @@ private function getUpdateStartTime(): \DateTimeImmutable
+     * @param string $target
+     *   The path of the target file. Needs to be known to the most recent
+     *   targets metadata downloaded in ::refresh().
+     * @param string[] $searchedRoles
+     *   The roles that have already been searched. This is for internal use only and should not be passed by calling code.
+     *   calls to this function and should be provided by any callers.
+     * @param bool $terminated
+     *
+     * @return \Tuf\Metadata\TargetsMetadata|null
+     *   The target metadata that is the end node for search or null if the target metadata does not make any delegations.
+     *   The target metadata or may not contain information for the target.
+     */
+    private function performDelegationSearch(TargetsMetadata $targetsMetadata, string $target, array $searchedRoles, bool &$terminated = false): ?TargetsMetadata
+    {
+        foreach ($targetsMetadata->getDelegatedKeys() as $keyId => $delegatedKey) {
+            $this->signatureVerifier->addKey($keyId, $delegatedKey);
+        }
+        foreach ($targetsMetadata->getDelegatedRoles() as $delegatedRole) {
+            $delegatedRoleName = $delegatedRole->getName();
+            if (in_array($delegatedRoleName, $searchedRoles, true)) {
+                // § 5.6.7.1
+                // If this role has been visited before, then skip this role (so that cycles in the delegation graph are avoided).
+                continue;
+            }
+            if (count($searchedRoles) > static::MAXIMUM_TARGET_ROLES) {
+                return $delegatedTargetsMetadata ?? null;
+            }
+
+            $this->signatureVerifier->addRole($delegatedRole);
+            if ($delegatedRole->matchesPath($target)) {
+                // Targets must match the path in all roles in the delegation chain, so if the path does not match,
+                // do not evaluate this role or any roles it delegates to.
+                if ($delegatedRole->isTerminating()) {
+                    $terminated = true;
+                }
+
+                $this->fetchAndVerifyTargetsMetadata($delegatedRoleName);
+                /** @var \Tuf\Metadata\TargetsMetadata $delegatedTargetsMetadata */
+                $delegatedTargetsMetadata = $this->metadataFactory->load($delegatedRoleName);
+                if ($delegatedTargetsMetadata->hasTarget($target)) {
+                    return $delegatedTargetsMetadata;
+                }
+                $searchedRoles[] = $delegatedRoleName;
+                // § 5.6.7.2.1
+                // Recursively search the list of delegations in order of appearance.
+                if ($resolvedTargetMetadata = $this->performDelegationSearch($delegatedTargetsMetadata, $target, $searchedRoles, $terminated)) {
+                    if ($terminated || $resolvedTargetMetadata->hasTarget($target)) {
+                        return $resolvedTargetMetadata;
+                    }
+                }
+
+                if ($delegatedRole->isTerminating()) {
```

Comment:
```suggestion
                if ($terminated) {
```

Reply:
This actually will cause tests to fail. 

If you look at the case `TUFTestFixtureTerminatingDelegation f.txt` this will fail because F will be found when it should not be found. Same case with `TUFTestFixtureNestedTerminatingNonDelegatingDelegation d.txt` this should throw a not found.

We actually don't need 

```
 if ($delegatedRole->isTerminating()) {
       $terminated = true;
 }
```
Before our recursive call to `performDelegationSearch()`. I moved  it after.

I can update comments in code later to explain why this is the case. Sorry don't have time right now.

Reply:
Okay, no problem. I agree that more comments will be useful.

Reply:
I updated this logic to make it simpler. Take another look

Reply:
@phenaproxima fyi part of the simplification here is that `searchDelegatedRolesForTarget()` only returns either the metadata contains the target info or null if no result. Before it could also return last metadata searched regardless of whether the target was found.